loki-logback-appender/src/main/java/com/github/loki4j/logback/JsonLayout.java

```diff
@@ -48,7 +48,7 @@ public class JsonLayout extends ContextAwareBase implements Layout<ILoggingEvent
     private List<JsonProvider<ILoggingEvent>> customProviders = new ArrayList<>();
 
     @Override
-    public String doLayout(ILoggingEvent event) {
+    public synchronized String doLayout(ILoggingEvent event) {
```

Comment:
I guess `ThreadLocal<JsonEventWriter>` would be a better alternative for `synchronized` in terms of performance. Could you please fix this?

Reply:
I thought about using a ThreadLocal, but decided against it. A ThreadLocal would create a buffer for every thread that ever writes a log. Since we cannot know when the last log is written, we cannot clean up the threadlocal. This becomes even more problematic, when not using pooled threads or when using virtual threads – using a thread local in this case is basically a memory leak by design.

I have not measured, but I doubt that `synchronized` is causing a noticable slowdown. However, if you are worried about concurrency when multiple threads write logs (at the very same time), I would instead propose to use the other straightforward solution: initialize a new `JsonEventWriter` instance every time a logging event is layouted. The instance and its buffer only live for the duration of the method call and the memory will be freed immediately by any sensible JVM implementation.

Only when we have identified either `synchronized` or allocating new JsonEventWriter instances as bottleneck, I would rewrite to a more sophisticated solution, such as using a bounded writer/buffer pool.

If you think this is non-negiotable, I'll gladly implement the pooled version now.

Reply:
Thanks for such a detailed reply! Interesting stuff. So looks like we have the following options:

1. synchronized
2. ThreadLocal
3. allocate new `JsonEventWriter` per msg
4. pool of `JsonEventWriter`

I'm open to any possible solution. Options 1-3 are really easy to implement and check the performance. I have a very basic benchmark engine ([layout benchmark](https://github.com/loki4j/loki-logback-appender/blob/main/loki-logback-appender/src/test/java/com/github/loki4j/logback/performance/LayoutsTest.java), [multi-thread benchmark](https://github.com/loki4j/loki-logback-appender/blob/main/loki-logback-appender/src/test/java/com/github/loki4j/logback/performance/AppenderTest.java#L60)). So if you can contribute tests for these options, that would be interesting to see!

Also some side notes:

> Since we cannot know when the last log is written, we cannot clean up the threadlocal. 

GC collects thread locals when the thread is terminated. So it'll be a problem only if threads in user's app are leaking, but that's not something we as a logging lib are supposed to deal with.

>  when using virtual threads – using a thread local in this case is basically a memory leak by design

True, but `synchronized` is also not welcomed for virtual threads (see https://blog.ycrash.io/pitfalls-to-avoid-when-switching-to-virtual-threads/). Honestly, I'm not sure we should strive to support virtual threads before we have industrial best practices / solutions in place. Maybe some are already there, I haven't yet dug deep into this topic.

Reply:
Good to know that `synchronized` doesn't work well with virtual threads. You are right about ThreadLocals too, but that can still leave a lot of buffers lying around if you have a mid-sized application with large thread pools. Nevertheless, I have extended the layout benchmarks and executed it against the 3 options (and the baseline, i.e. the functionally broken version from master). I set the parallelism to 4 (I have a 4-core CPU with 8 threads) and added 3 benchmarks for the same layout. I haven't measured memory usage or GC times. I'm curious though, why the benchmarks aren't written with something like JMH :) Anyhow, here are the results:

Baseline (the buggy version: a single buffer shared by multiple threads):

```
Run: jsonLayout1 #0, duration = 29437.68 ms, throughput = 339,700.644 rps, eff duration = 8007.09 ms, eff throughput = 1,248,892.685 rps
Run: jsonLayout2 #0, duration = 28929.28 ms, throughput = 345,670.495 rps, eff duration = 7863.87 ms, eff throughput = 1,271,638.788 rps
Run: jsonLayout3 #0, duration = 28766.70 ms, throughput = 347,624.188 rps, eff duration = 7872.82 ms, eff throughput = 1,270,193.077 rps
```

Method-local writers (allocate a new writer instance for each method call):

```
Run: jsonLayout1 #0, duration = 22130.61 ms, throughput = 451,862.814 rps, eff duration = 6242.10 ms, eff throughput = 1,602,025.875 rps
Run: jsonLayout2 #0, duration = 21646.17 ms, throughput = 461,975.464 rps, eff duration = 6099.97 ms, eff throughput = 1,639,351.351 rps
Run: jsonLayout3 #0, duration = 21499.05 ms, throughput = 465,136.784 rps, eff duration = 6077.12 ms, eff throughput = 1,645,516.751 rps

Synchronized (simply marking the method as `synchronized`):

```
Run: jsonLayout1 #0, duration = 70868.90 ms, throughput = 141,105.625 rps, eff duration = 18409.36 ms, eff throughput = 543,201.790 rps
Run: jsonLayout2 #0, duration = 73637.38 ms, throughput = 135,800.590 rps, eff duration = 19165.92 ms, eff throughput = 521,759.502 rps
Run: jsonLayout3 #0, duration = 72545.57 ms, throughput = 137,844.390 rps, eff duration = 18780.08 ms, eff throughput = 532,479.081 rps
```

ThreadLocal (having a single ThreadLocal field with a separate writer instance per thread):

```
Run: jsonLayout1 #0, duration = 17509.19 ms, throughput = 571,128.747 rps, eff duration = 5056.65 ms, eff throughput = 1,977,593.612 rps
Run: jsonLayout2 #0, duration = 16961.89 ms, throughput = 589,556.777 rps, eff duration = 4858.87 ms, eff throughput = 2,058,091.020 rps
Run: jsonLayout3 #0, duration = 21120.30 ms, throughput = 473,478.186 rps, eff duration = 6109.15 ms, eff throughput = 1,636,888.792 rps
```

Ordered by throughput we seem to get the following list:

1. ThreadLocal: 544 k rps, eff 1890 k rps (fastest)
2. Method-local writers: 459 k rps, eff 1629 k rps
3. Baseline: 344 k rps, eff 1264 k rps
4. Synchronized: 138 k rps, eff 532 k rps (slowest)

Interestingly, allocating a new writer for every method call seems to be faster than sharing a single writer instance. I don't have an explanation for that.

I have not run the tests with different parallelism configurations, so I cannot say anything about high-contention scenarios with many threads calling `doLayout` at the same time.

I'd still lean towards separate writer instances, because they are only slightly slower, the code is more straightforward, and the buffers are not kept around when large thread pools are in use. But it is your library, so let me know which variant you prefer!

Reply:
> I'd still lean towards separate writer instances, because they are only slightly slower, the code is more straightforward, and the buffers are not kept around when large thread pools are in use.

Agree, let's do new writers.

> I'm curious though, why the benchmarks aren't written with something like JMH

Well, I'm not sure if it has any benefits for my case, but it adds complexity for sure.